sql/src/main/java/io/crate/operation/projectors/ShardDMLExecutor.java

```diff
@@ -0,0 +1,200 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.projectors;
+
+import com.carrotsearch.hppc.IntArrayList;
+import io.crate.action.FutureActionListener;
+import io.crate.action.LimitedExponentialBackoff;
+import io.crate.data.BatchIterator;
+import io.crate.data.Row;
+import io.crate.data.Row1;
+import io.crate.executor.transport.ShardRequest;
+import io.crate.executor.transport.ShardResponse;
+import io.crate.operation.NodeJobsCounter;
+import io.crate.operation.collect.CollectExpression;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.bulk.BackoffPolicy;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.logging.Loggers;
+
+import javax.annotation.Nullable;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.BooleanSupplier;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static io.crate.operation.NodeJobsCounter.MAX_NODE_CONCURRENT_OPERATIONS;
+
+public class ShardDMLExecutor<TReq extends ShardRequest<TReq, TItem>, TItem extends ShardRequest.Item>
+    implements Function<BatchIterator, CompletableFuture<? extends Iterable<Row>>> {
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardDMLExecutor.class);
+
+    private static final BackoffPolicy BACKOFF_POLICY = LimitedExponentialBackoff.limitedExponential(1000);
+    public static final int DEFAULT_BULK_SIZE = 10_000;
+
+    private final int bulkSize;
+    private final ScheduledExecutorService scheduler;
+    private final BitSet responses;
+    private final ClusterService clusterService;
+    private final NodeJobsCounter nodeJobsCounter;
+    private final AtomicInteger pendingItemsCount = new AtomicInteger(0);
+    private final Consumer<Row> rowConsumer;
+    private final BooleanSupplier backpressureTrigger;
+    private final Function<Boolean, CompletableFuture<BitSet>> execute;
+
+    private TReq currentRequest;
+    private int numItems = -1;
+    private CompletableFuture executionFuture = new CompletableFuture();
+
+    public ShardDMLExecutor(int bulkSize,
+                            ScheduledExecutorService scheduler,
+                            CollectExpression<Row, ?> uidExpression,
```

Comment:
That's coming from the InputFactory (which is something we might want to make more specific at some point, but it's definitely out of scope for this PR)